File: python/shortest_bridge.py
from collections import deque


class Solution:
    def shortestBridge(self, grid: list[list[int]]) -> int:
        """time complexity: O(n^2), where n is the length of the grid"""
        visited: set[tuple[int, int]] = set()
        dirs: list[tuple[int, int]] = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        for r in range(len(grid)):
            for c in range(len(grid[r])):
                if grid[r][c] == 1:
                    self.dfs(r, c, grid, dirs, visited)
                    return self.bfs(grid, dirs, visited)
        return -1

    def dfs(
        self,
        row: int,
        col: int,
        grid: list[list[int, int]],
        dirs: list[tuple[int, int]],
        visited: set[tuple[int, int]],
    ):
        if (
            min(row, col) < 0
            or max(row, col) >= len(grid)
            or grid[row][col] == 0
            or (row, col) in visited
        ):
            return
        visited.add((row, col))
        for dir_r, dir_c in dirs:
            self.dfs(row + dir_r, col + dir_c, grid, dirs, visited)

    def bfs(
        self,
        grid: list[list[int, int]],
        dirs: list[tuple[int, int]],
        visited: set[tuple[int, int]],
    ) -> int:
        queue: deque[tuple[int, int]] = deque(visited)
        result: int = 0
        while len(queue) > 0:
            for _ in range(len(queue)):
                row, col = queue.popleft()
                for dir_r, dir_c in dirs:
                    next_r: int = row + dir_r
                    next_c: int = col + dir_c
                    if (
                        min(next_r, next_c) < 0
                        or max(next_r, next_c) >= len(grid)
                        or (next_r, next_c) in visited
                    ):
                        continue
                    if grid[next_r][next_c] == 1:
                        return result
                    queue.append((next_r, next_c))
                    visited.add((next_r, next_c))
            result += 1
        return -1


def test_shortestBridge_case_1():
    # arrange
    grid: list[list[int]] = [[0, 1], [1, 0]]
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.shortestBridge(grid)

    # assert
    assert expected == actual


def test_shortestBridge_case_2():
    # arrange
    grid: list[list[int]] = [[0, 1, 0], [0, 0, 0], [0, 0, 1]]
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.shortestBridge(grid)

    # assert
    assert expected == actual


def test_shortestBridge_case_3():
    # arrange
    grid: list[list[int]] = [
        [1, 1, 1, 1, 1],
        [1, 0, 0, 0, 1],
        [1, 0, 1, 0, 1],
        [1, 0, 0, 0, 1],
        [1, 1, 1, 1, 1],
    ]
    expected: int = 1

    # act
    solution = Solution()
    actual = solution.shortestBridge(grid)

    # assert
    assert expected == actual

File: rust/src/code/shortest_bridge.rs
use std::cmp::{max, min};
use std::collections::{HashSet, VecDeque};

struct Solution;

impl Solution {
    pub fn shortest_bridge(grid: Vec<Vec<i32>>) -> i32 {
        let mut visited: HashSet<(i32, i32)> = HashSet::new();
        let dirs: Vec<(i32, i32)> = vec![(1, 0), (-1, 0), (0, 1), (0, -1)];
        for r in 0..grid.len() {
            for c in 0..grid[r].len() {
                if grid[r][c] == 1 {
                    Self::dfs(r as i32, c as i32, &dirs, &grid, &mut visited);
                    return Self::bfs(&dirs, &grid, &mut visited);
                }
            }
        }
        -1
    }

    fn dfs(
        row: i32,
        col: i32,
        dirs: &Vec<(i32, i32)>,
        grid: &Vec<Vec<i32>>,
        visited: &mut HashSet<(i32, i32)>,
    ) {
        if min(row, col) < 0
            || max(row, col) >= grid.len() as i32
            || visited.contains(&(row, col))
            || grid[row as usize][col as usize] == 0
        {
            return;
        }
        visited.insert((row, col));
        for (dir_r, dir_c) in dirs.iter() {
            Self::dfs(row + dir_r, col + dir_c, dirs, grid, visited);
        }
    }

    fn bfs(dirs: &Vec<(i32, i32)>, grid: &Vec<Vec<i32>>, visited: &mut HashSet<(i32, i32)>) -> i32 {
        let mut queue: VecDeque<(i32, i32)> = VecDeque::new();
        let mut result: i32 = 0;
        for (r, c) in visited.iter() {
            queue.push_back((*r, *c));
        }
        while queue.len() > 0 {
            for _ in 0..queue.len() {
                if let Some((row, col)) = queue.pop_front() {
                    for (dir_r, dir_c) in dirs {
                        let next_r: i32 = row + dir_r;
                        let next_c: i32 = col + dir_c;
                        if min(next_r, next_c) < 0
                            || max(next_r, next_c) >= grid.len() as i32
                            || visited.contains(&(next_r, next_c))
                        {
                            continue;
                        }
                        if grid[next_r as usize][next_c as usize] == 1 {
                            return result;
                        }
                        queue.push_back((next_r, next_c));
                        visited.insert((next_r, next_c));
                    }
                }
            }
            result += 1;
        }

        -1
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn test_shortest_bridge_case_1() {
        // arrange
        let grid: Vec<Vec<i32>> = vec![vec![0, 1], vec![1, 0]];
        let expected: i32 = 1;

        // act
        let mut actual = Solution::shortest_bridge(grid);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_shortest_bridge_case_2() {
        // arrange
        let grid: Vec<Vec<i32>> = vec![vec![0, 1, 0], vec![0, 0, 0], vec![0, 0, 1]];
        let expected: i32 = 2;

        // act
        let mut actual = Solution::shortest_bridge(grid);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_shortest_bridge_case_3() {
        // arrange
        let grid: Vec<Vec<i32>> = vec![
            vec![1, 1, 1, 1, 1],
            vec![1, 0, 0, 0, 1],
            vec![1, 0, 1, 0, 1],
            vec![1, 0, 0, 0, 1],
            vec![1, 1, 1, 1, 1],
        ];
        let expected: i32 = 1;

        // act
        let mut actual = Solution::shortest_bridge(grid);

        // assert
        assert_eq!(expected, actual);
    }
}
